doc/test-framework/writing-e2e-tests.md

```diff
@@ -0,0 +1,209 @@
+The next step is to create a TestCtx for the current test and defer its cleanup function:
+```
+ctx := framework.NewTestCtx(t)
+defer ctx.Cleanup(t)
+```
+
+Now that there is a TestCtx, the test's kubernetes resources (specifically the RBAC and Operator deployment)
+can be initialized:
+```
+err := ctx.InitializeClusterResources()
+if err != nil {
+    t.Fatalf("failed to initialize cluster resources: %v", err)
+}
+```
+
+If you want to make sure the operator's deployment is fully ready before moving onto the next part of the
+test, the `WaitForDeployment` function from e2eutil (in the sdk under `pkg/test/e2eutil`) can be used:
+```
+// get namespace
+namespace, err := ctx.GetNamespace()
+if err != nil {
+    t.Fatal(err)
+}
+// get global framework variables
+f := framework.Global
+// wait for memcached-operator to be ready
+err = e2eutil.WaitForDeployment(t, f.KubeClient, namespace, "memcached-operator", 1, time.Second*5, time.Second*30)
+if err != nil {
+    t.Fatal(err)
+}
+```
+
+#### 4. Write the test specific code
+Now that the operator is ready, we can create a custom resource. Since the controller-runtime's dynamic client uses
+go contexts, make sure to import the go context library. In this example, we imported it as `goctx`:
+```
+// create memcached custom resource
+exampleMemcached := &cachev1alpha1.Memcached{
+    TypeMeta: metav1.TypeMeta{
+        Kind:       "Memcached",
+        APIVersion: "cache.example.com/v1alpha1",
+    },
+    ObjectMeta: metav1.ObjectMeta{
+        Name:      "example-memcached",
+        Namespace: namespace,
+    },
+    Spec: cachev1alpha1.MemcachedSpec{
+        Size: 3,
+    },
+}
+err = f.DynamicClient.Create(goctx.TODO(), exampleMemcached)
+if err != nil {
+    return err
+}
+```
+
+Now we can check if the operator successfully worked. In the case of the memcached operator, it should have
+created a deployment called "example-memcached" with 3 replicas:
+```
+// wait for example-memcached to reach 3 replicas
+err = e2eutil.WaitForDeployment(t, f.KubeClient, namespace, "example-memcached", 3, time.Second*5, time.Second*30)
+if err != nil {
+    return err
+}
+```
+ 
+We can also test that the deployment scales correctly when the CR is updated:
+```
+err = f.DynamicClient.Get(goctx.TODO(), types.NamespacedName{Name: "example-memcached", Namespace: namespace}, exampleMemcached)
+if err != nil {
+    return err
+}
+exampleMemcached.Spec.Size = 4
+err = f.DynamicClient.Update(goctx.TODO(), exampleMemcached)
+if err != nil {
+    return err
+}
+
+// wait for example-memcached to reach 4 replicas
+err = e2eutil.WaitForDeployment(t, f.KubeClient, namespace, "example-memcached", 4, time.Second*5, time.Second*30)
+if err != nil {
+    return err
+}
+```
+
+Now that the test is complete, we are done with the function. Once the end of the function is reached, the TestCtx's cleanup
+functions will automatically be run since they were deferred when the TestCtx was created.
```

Comment:
we probably want to link the actual code from https://github.com/operator-framework/operator-sdk-samples/tree/master/memcached-operator/test/e2e to there.

Reply:
We should link it at the top. When we start the section Individual Tests we should mention we're using https://github.com/operator-framework/operator-sdk-samples/blob/master/memcached-operator/test/e2e/memcached_test.go as an example.